lang/src/org/partiql/lang/eval/physical/window/NavigationWindowFunction.kt

```diff
@@ -23,13 +25,16 @@ abstract class NavigationWindowFunction() : WindowFunction {
     override fun processRow(
         state: EvaluatorState,
         arguments: List<ValueExpression>,
-        windowVarDecl: SetVariableFunc
+        windowVarDecl: PartiqlPhysical.VarDecl
     ) {
         state.load(currentPartition[currentPos])
         val value = processRow(state, arguments, currentPos)
         // before we declare the window function result, we need to go back to the current row
         state.load(currentPartition[currentPos])
-        windowVarDecl(state, value)
+        windowVarDecl.toSetVariableFunc().invoke(state, value)
+        // make sure the change of state is reflected in the partition
+        // so the result of the current window function won't get removed by the time we process the next window function at the same row level.
+        currentPartition[currentPos][windowVarDecl.index.value.toInt()] = state.registers[windowVarDecl.index.value.toInt()]
```

Comment:
I left a note within `WindowFunction` regarding this exact line. If we were to go about creating the `partitions` to be a list of `EvaluatorState`, then the suggestion would actually be:
```suggestion
        setWindowVar.invoke(state, value)
        setWindowVar.invoke(currentPartition[currentPos], value)
```

Assuming `setWindowVar` is a `SetVariableFunc`, and `currentPartition[currentPos]` is an `EvaluatorState`.

Reply:
Love the comment here. I actually modeled a few different way:
1. Each partition is a array of `register`, and we use SetVariableFunc:
    Doing so we loss the ability to get the index of the global register the result of the given window function should go to, the solution will be to loop and rewrite the state's register. Now if the state contains a large number of element, the overhead is large. 
2. Each partition is a array of `register` and we use `VarDecl`:  (This is the current approach)
    As you pointed out, this way is not uniform with other operator, and directly access and modifying the "register alike data" is indeed an unsafe operation in that, if there were to be a bug, it would be difficult to debug... But this does reduce the overhead. 

Please correct me if I am wrong but you seem to be suggesting a third way: which is 
Each partition is a array of `EvalutorState` and we use `SetVariableFunc`:
I have not yet modeled this approach, but my first concern is that if we create a evaluatorState for essentially each row in the table, the overhead is quite large because for each of those we have to carry a reference of session data.

Reply:
Yeah, I'm not excited about the state either. Just wanted to bring it up as something the other operators have used. Anyways, since we have yet to fully decide on how we're handling state, we'll leave this as a TODO to come back to.